### PR TITLE
Label the scatterplot 

### LINKED ISSUE #355: Response Correlation Viewer: Display which response and parameter are plotted in the scatterplot
**Is your feature request related to a problem? Please describe.**
Looking at the scatter plot right now, it's a bit hard to parse. There are no labels besides the scales of the axes.

**Describe the solution you'd like**
We woud like to be able to see
- correlation between which response and parameter is displayed, at which index

**Describe alternatives you've considered**
Just looking at the bottom of the page, to the info text, is a bit far from the plot and seems to lead to bad user experience.

### PR DESCRIPTION
**Pull request by valentin-krasontovitsch**
**Issue**
Closes #355

**Approach**
We add titles, and fix hover texts.

In a silly way, some options available in the plotly API did not have any effect when I tried to use them, for instance `x_title` should add a title at the bottom of the scatterplot/histogram combination, but, as I said, no effect : / so it's on the side!

## Pre review checklist

- [X] Added appropriate labels

I did this on top of the work of #354, so we should wait for that one to be merged! making this a draft for now.

### CODE REVIEW

**Comment by valentin-krasontovitsch** [comment]
sorry forgot to add screenshots, here we go - before the changes:

scatterplot-labels-before-mini
after the changes:

scatterplot-labels-after-mini

File: webviz_ert/controllers/response_correlation_controller.py [hunk 1]
```diff
@@ -551,6 +568,16 @@ def _get_selected_ensembles_from_store(
...
+        if isinstance(raw_value, str):
+            try:
+                datetime_value = pd.to_datetime(raw_value)
+                return datetime_value.date()
+            except (pd.errors.ParserError):
+                pass
+        return raw_value
+
 
 def sort_dataframe(
```
Comment (hnformentin) [inline webviz_ert/controllers/response_correlation_controller.py:582]: Should it be internal `_`?

Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:582]: the `sort_dataframe` function? outside the scope of this PR, but following the campsite rule, i guess we can make it internal. i tried to make it reusable, but the part where you check if the index exists, that's probably very particular to our case.

**Comment by valentin-krasontovitsch** [comment]
tooltips have also been improved - before:

scatterplot-labels-tooltips-before

after:

scatterplot-labels-tooltips-after

disclaimer: images were photoshopped to overlay two tooltips in one image :innocent:

File: webviz_ert/controllers/response_correlation_controller.py [hunk 2]
```diff
@@ -551,6 +568,16 @@ def _get_selected_ensembles_from_store(
             ]
         return None
 
+    def _format_index_value(axis: pd.Index, index: int) -> Union[int, datetime.date]:
```
Comment (hnformentin) [inline webviz_ert/controllers/response_correlation_controller.py:571]: I think that it would be more consistent with the other functions if we had them all after `controller`. What do you think?

Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:571]: yes sorry, of course! done.

File: webviz_ert/controllers/response_correlation_controller.py (hunk 1 above)
Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py]: done

**Comment by hnformentin** [comment]
Testing the functionality...a surprising behaviour for me was that the name of the response and parameter was not in the axis, but in the title of the scatterplot. As a user I would expect to have these names in the positions that I tried to represent in the image below :D 
Screenshot 2022-08-04 at 14 44 32

Do you think that my surprise adds something, or are you surprised with my surprise?

**Comment by hnformentin** [comment]
But in any case, the proposal is already better than not having any indication of what was parameter and what was response 🎉

File: tests/controllers/test_response_correlation_controller.py [hunk 3]
```diff
@@ -8,19 +8,19 @@
 )
 
 from webviz_ert.controllers.response_correlation_controller import (
-    sort_dataframe,
+    _sort_dataframe,
 )
 
 
-def test_sort_dataframe():
+def test__sort_dataframe():
```
Comment (hnformentin) [inline tests/controllers/test_response_correlation_controller.py:15]: Different than the tests for the other internal functions, where I (not sure if we) used only one `_`. Do you think we should change all of them to be consistent though the test file?

File: webviz_ert/controllers/response_correlation_controller.py [hunk 4]
```diff
@@ -322,14 +323,16 @@ def update_response_parameter_scatterplot(
                         PlotModel(
                             x_axis=x_data.values.flatten(),
                             y_axis=y_data.values.flatten(),
-                            text="Mean",
-                            name=f"{repr(ensemble)}: {active_response}x{active_parameter}@{response.axis[x_index]}",
+                            text="Mean",  # TODO ask about this!
```
Comment (hnformentin) [inline webviz_ert/controllers/response_correlation_controller.py:326]: `# TODO ask about this!` what is this comment about?

File: webviz_ert/controllers/response_correlation_controller.py [hunk 5]
```diff
@@ -610,3 +627,14 @@ def _get_first_observation_x(obs_data: pd.DataFrame) -> Union[int, str]:
...
+
+def _format_index_value(axis: pd.Index, index: int) -> Union[int, datetime.date]:
+    raw_value = axis[index]
+    if isinstance(raw_value, str):
+        try:
+            datetime_value = pd.to_datetime(raw_value)
+            return datetime_value.date()
+        except (pd.errors.ParserError):
+            pass
+    return raw_value
```
Comment (hnformentin) [inline webviz_ert/controllers/response_correlation_controller.py:640]: This function could be tested. It would be good to document which states we expect to have a proper value and which states the exception is to be raised.

**Comment by hnformentin** [comment]
> **Issue** Closes #355
> 
> **Approach** We add titles, and fix hover texts.
> 
> In a silly way, some options available in the plotly API did not have any effect when I tried to use them, for instance `x_title` should add a title at the bottom of the scatterplot/histogram combination, but, as I said, no effect : / so it's on the side!
> 
> ## Pre review checklist
> * [x]  Added appropriate labels
> 
> I did this on top of the work of #354, so we should wait for that one to be merged! making this a draft for now.

I was able to have proper titles on the axis of the scatterplot. Just added:
`
fig.update_layout({"xaxis": {"title": {"text": f"{active_response} @ {formatted_x_value}"}}})
fig.update_layout({"yaxis": {"title": {"text": f"{active_parameter}"}}})
`
after
`
fig.update_layout(assets.ERTSTYLE["figure"]["layout"])
fig.update_layout(showlegend=False)
`

and got 

Screenshot 2022-08-05 at 09 03 38

Now for the histogram it is another problem :-) I am trying to see where the layout properties of the histograms are, but so far did not find it. Any idea?

File: tests/controllers/test_response_correlation_controller.py (hunk 3 above)
Comment (valentin-krasontovitsch) [reply tests/controllers/test_response_correlation_controller.py:15]: ah, this was just a result of blind replacing / renaming. gonna fix it in this instance, don't think we need a double underscore, it's just a name after all : )

and done.

File: webviz_ert/controllers/response_correlation_controller.py (hunk 4 above)
Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:326]: XD that's hilarious, maybe they put it to have something? my comment is indeed for finding out why this is called mean, and i also thought that it didn't make sense, but since i don't know much about the domain or statistics, i thought i'd ask ^^

ok i figured out that no plot was drawn because the plot model wants to get a name. extending the plot model to accept no given name, we fix the no plot and circumvent having empty space / an empty line in the tooltip. yeahy : )

File: webviz_ert/controllers/response_correlation_controller.py (hunk 5 above)
Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:640]: thanks for pointing that out - I wrote a parametrized test, and found an error (or not considered case) in the code, so that was nice ^^

have a look and see if it's alright please : )

**Comment by valentin-krasontovitsch** [comment]
i am completely confused about the roles of layout and non-layout properties, but it seems like I found a way of settings axes titles (found in [plotly examples about subplots](https://plotly.com/python/subplots/) under **Customizing Subplot Axes**)

pushed the changes, observe that there's some potentially silly half-DRY business going on with storing the number of rows in a variable.

Also please observe that I have made space between the subplots, which makes the scatterplot  a bit smaller.

here's the changes isolated

new-titles-mini

and here it is in context

new-titles-in-context

I might try to play with it, to decrease the size of the spacing a bit more!

but for now - what do you think, Helena?

File: tests/controllers/test_response_correlation_controller.py [hunk 6]
```diff
@@ -114,3 +115,18 @@ def test_get_first_observation_x_invalid():
...
+@pytest.mark.parametrize(
+    "raw_value,expected_formatted_value",
+    [
+        ("2022-08-05 14:25:00", "2022-08-05"),
+        ("2022-09-21 14:25:00", "2022-09-21"),
+        (14, "14"),
+        ("213", "213"),
+        ("SPAM", "SPAM"),
+    ],
+)
```
Comment (hnformentin) [inline tests/controllers/test_response_correlation_controller.py:129]: Very nice, I would just suggest the order of the tests and rename one of them. In this way, the test cases are clearly expressing what is written in the string of the function:
```
@pytest.mark.parametrize(
    "raw_value,expected_formatted_value",
    [
        ("2022-08-05 14:25:00", "2022-08-05"),
        ("2022-09-21 14:25:00", "2022-09-21"),
        ("PARSING_FAILS_FROM_NOW", "PARSING_FAILS_FROM_NOW"),
        (14, "14"),
        ("213", "213"),
    ],
)
```

**Comment by hnformentin** [comment]
> 

That is really nice @valentin-krasontovitsch !
You managed to add all these labels. I am curious about how you implemented it 🤓 

Ultimately, it would be better to ask for user feedback directly, nevertheless, some considerations that I would like to share.

If I were a user, I would be happy to have the labels on the axes of the plots, not necessarily the titles of them. Specially if I needed to choose between having larger plots or their titles. This seems to be a good point to ask feedback, and since we already there with the implementation of the titles, it would be cool to let the code related to the titles/size flexible (if possible) and ask for feedback later when they are back.

It is really nice to have quick access to all the ensembles selected 👏 Nevertheless, if I were a user, I would prefer to analyse the scatterplot and histograms of each ensemble in a plot separately. This is because when we look for these kind of plots, we try to capture trends within the ensemble data and outliers. 

Some context about data analysis, if you are interested:
- With the scatterplot we can observe trends that diverge from linear correlation (since the correlation coefficient that we calculate represents only linear correlation...imagine that a synodal correlation would correspond to a zero linear correlation, displayed in the heat map) and outliers, which are points that deform the calculated correlation, and maybe not representative of the phenomena been modelled or even lead to investigate a new phenomena that we were not aware about.
- With the histograms we can observe trends that diverge from normal distribution (since many statistical methods are based on this assumption) and outliers.

Having multiple ensembles plotted at the same time prevent me to see the trends and outliers within an ensemble. At the same time, having the possibility of displaying  different ensembles is a nice feature to analyse changes between two or several ensembles. But we can do this by clicking in the heat map at same parameters x response combination, but different ensembles. If this becomes too difficult because of a larger number of parameters and responses (and consequently very small rectangles displayed on the heat map), I would prefer to selector buttons with which I could choose dynamically the ensemble to be plotted.

Does any of this make sense?

File: webviz_ert/controllers/response_correlation_controller.py [hunk 7]
```diff
@@ -610,3 +628,32 @@ def _get_first_observation_x(obs_data: pd.DataFrame) -> Union[int, str]:
...
+    active_parameter: str,
+    x_value: Union[datetime.date, np.number],
+    n_rows: int,
+) -> None:
+    fig.update_layout(assets.ERTSTYLE["figure"]["layout"])
+    fig.update_xaxes(title_text=f"{active_response} @ {x_value}", row=1, col=1)
+    fig.update_yaxes(title_text=f"{active_parameter}", row=1, col=1)
+    fig.update_xaxes(title_text=f"{active_parameter}", row=n_rows, col=1)
+    fig.update_xaxes(title_text=f"{active_response}", row=n_rows, col=2)
+    fig.update_layout(showlegend=False)
```
Comment (hnformentin) [inline webviz_ert/controllers/response_correlation_controller.py:659]: In other parts of the code, we have updated the layout via a dictionary and not passing the `fig` element, e.g. `_define_style_ensemble` or `_layout_figure`. This was chosen to have simpler variables been passed and consequently easier to test if we want so. What would you think about maintaining a similar pattern for layout functions?

Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:659]: are you referring just to the last line, 659, setting visibility of the legend? or also to the `update_?axes` calls?

if you mean the latter, I can try to see if this can be replicated by passing a dict to `update_layout`

File: tests/controllers/test_response_correlation_controller.py (hunk 6 above)
Comment (valentin-krasontovitsch) [reply tests/controllers/test_response_correlation_controller.py:129]: very good suggestion - it seems like we can actually add descriptive names to our test cases, cf. [the pytest parametrize docs on test IDs](https://docs.pytest.org/en/7.1.x/example/parametrize.html#different-options-for-test-ids)

i'm gonna try to have a go at that!

Comment (valentin-krasontovitsch) [reply tests/controllers/test_response_correlation_controller.py:129]: I want to insist a little bit on keeping the `SPAM` as a silly string, as it seems to be [the pythonic way](https://docs.python.org/3.10/library/pprint.html) (see first code block) - but it's not a hard insist : )

Comment (valentin-krasontovitsch) [reply tests/controllers/test_response_correlation_controller.py:129]: considering your suggestion of `PARSINGS_FAILS_FROM_NOW`, I would like to avoid adding that information ("parsing fails") to the test cases - i think it's an implementation details of the function. what the test wants to know, is: if I put in something that is not a date, will I get out the same thing, possibly converted to a string if it was a number.

does that sound OK? pushed the updated version with IDs

**Comment by valentin-krasontovitsch** [comment]
> That is really nice @valentin-krasontovitsch ! 

Thank you ^^

> You managed to add all these labels. I am curious about how you implemented it nerd_face

i know, right? :grin:  in some other docs / example i found the methods `update_xaxes` and `update_yaxes` used on the figure object in a multiplot context to achieve exactly what we wanted! just needed to google moar... a link is in [a comment above on this PR](https://github.com/equinor/webviz-ert/pull/356#issuecomment-1206513766)

> If I were a user, I would be happy to have the labels on the axes of the plots, not necessarily the titles of them. 

but the labels are on the axes now, aren't they? :thinking: see the [screenshots above, in this comment](https://github.com/equinor/webviz-ert/pull/356#issuecomment-1206513766)
i just added actual titles, like `Histogram Response`, but I can remove those if you think it's too much / superfluous information : ) very flexible there, no problem at all!

> It is really nice to have quick access to all the ensembles selected clap 

I'm sorry, but I cannot take credit for that - the multiple ensembles setup was already done before I started working on the labels (and are also not in the scope of this issue / PR).

If you think the current layout / display with multiple ensembles is not good, we could probably make an issue about it, and discuss with whoever came up with the design in the first place. 

> Some context about data analysis, if you are interested:
> 
> ...
> 
> Does any of this make sense?

Thank you, it definitely makes a lot of sense, and is very helpful ^^

When you explain it like that, I have to say that I agree and don't know why they designed it the way it is at the moment! Based on the git history, this seems to have been introduced in the original commit 194838f21ec81d82e9ab398e8ad4011c21d96540 that created the whole plugin, and @xjules was the author : )

Julius, do you have any comments? ^^

File: webviz_ert/controllers/response_correlation_controller.py (hunk 7 above)
Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:659]: i tried to take a look. following assumptions were made: using the `update_layout` function, we need to pass a dict (cf. [docs on `update_layout`)[https://plotly.com/python-api-reference/generated/plotly.graph_objects.Figure.html?highlight=update_layout#plotly.graph_objects.Figure.update_layout]. that dict should be of type [`plotly.graph_objects.layout`](https://plotly.github.io/plotly.py-docs/generated/plotly.graph_objects.Layout.html#plotly.graph_objects.Layout)

Now, that one has a property `Grid` where one can supply `xaxes`, but that one should be a list of strings matching some regex (`^x([2-9]|[1-9][0-9]+)?( domain)?$`), which among other things means "starts with x" - this does not seem to be what we need. also, the docs for that property mention

> Used with yaxes when the x and y axes are shared across columns and rows

so again this doesn't seem to be what we need, as we don't want to have axes shared across columns and rows

trying to set `xaxes` directly  in the layout dict (without `grid`) gives an error message.

it looks like the `update_?axes` methods cannot be replaced with a layout dict : /

File: tests/controllers/test_response_correlation_controller.py (hunk 6 above)
Comment (hnformentin) [reply tests/controllers/test_response_correlation_controller.py:130]: That is fine!

File: webviz_ert/controllers/response_correlation_controller.py (hunk 7 above)
Comment (hnformentin) [reply webviz_ert/controllers/response_correlation_controller.py:659]: I understand. Thank you for verifying that it is not possible and writing it out!

**Comment by hnformentin** [comment]
> i know, right? 😁 in some other docs / example i found the methods `update_xaxes` and `update_yaxes` used on the figure object in a multiplot context to achieve exactly what we wanted! just needed to google moar... a link is in [a comment above on this PR](https://github.com/equinor/webviz-ert/pull/356#issuecomment-1206513766)

Thanks for sharing it!

**Comment by hnformentin** [comment]
> but the labels are on the axes now, aren't they? 🤔 see the [screenshots above, in this comment](https://github.com/equinor/webviz-ert/pull/356#issuecomment-1206513766)
> i just added actual titles, like `Histogram Response`, but I can remove those if you think it's too much / superfluous information : ) very flexible there, no problem at all!

Yes, I thought was too much to have these titles because I would see better if there was more space to the plots. But it is up to you!

**Comment by hnformentin** [comment]
> I'm sorry, but I cannot take credit for that - the multiple ensembles setup was already done before I started working on the labels (and are also not in the scope of this issue / PR).
> 
> If you think the current layout / display with multiple ensembles is not good, we could probably make an issue about it, and discuss with whoever came up with the design in the first place.
> 
> > Some context about data analysis, if you are interested:
> > ...
> > Does any of this make sense?
> 
> Thank you, it definitely makes a lot of sense, and is very helpful ^^
> 
> When you explain it like that, I have to say that I agree and don't know why they designed it the way it is at the moment! Based on the git history, this seems to have been introduced in the original commit [194838f](https://github.com/equinor/webviz-ert/commit/194838f21ec81d82e9ab398e8ad4011c21d96540) that created the whole plugin, and @xjules was the author : )
> 
> Julius, do you have any comments? ^^

Indeed I did not notice we had that before!!! Thank you for doing the search.
And then I am curious why we have that since the beginning...

**Review by hnformentin — APPROVED** [review]

**Comment by valentin-krasontovitsch** [comment]
> Yes, I thought was too much to have these titles because I would see better if there was more space to the plots. But it is up to you!

right, sorry, I said I would try to make the plots bigger, but forgot :sweat_smile: 
now I tried it out - just doubling the number of rows, and scaling the plots accordingly, effectively shrinking the empty row used as space for the axes labels and titles.

here's the result, once with titles, once without

squeezed-with-titles

squeezed-without-titles

~~i would say the space is needed in any case for the axes labels~~ looking at the screenshots, its apparent that the plots become bigger without titles,so I'm gonna drop them for now, and other plots don't have no titles either, so that will be more consistent, too.

i have on my todo to make an issue about adding more visual hints in the response correlation viewer - titles for all plots could be part of that : )

**Comment by valentin-krasontovitsch** [comment]
.... and i forgot to adjust one index, which led to the disappearing of the x-axes labels of the bottom plots - gonna fix that.

File: webviz_ert/controllers/response_correlation_controller.py (hunk 7 above)
Comment (valentin-krasontovitsch) [reply webviz_ert/controllers/response_correlation_controller.py:662]: no problem! : )

**Comment by valentin-krasontovitsch** [comment]
OK, I've refactored a bit to make the numbers games I'm playing with rows in the multiplot a bit more DRY, and pulled stuff out into helper function.

Will you please take a look at the last commit, @hnformentin?

Latest screenshot:

final-scatterplot

**Comment by valentin-krasontovitsch** [comment]
Re: multiple ensembles - according to @xjules the current layout was requested from the previous product owner. he agrees that it feels cluttered at the moment. he advises further to wait for @oysteoh who might have comments on this particular aspect from actual users.

**Comment by valentin-krasontovitsch** [comment]
had to quickly fix a flaky test, that's that last commit from just now, sorry. hope it's OK to take it in this PR as it's a quick fix, using the known solution : )

**Comment by valentin-krasontovitsch** [comment]
and another commit, this time for fixing the dash version so that the test rewrite for fixing the flakiness works in jenkins

File: test_requirements.txt [hunk 8]
```diff
@@ -4,7 +4,7 @@ pytest-mock
 pylint
 black
 scipy
-dash[testing]
+dash[testing]==2.5.1
```
Comment (hnformentin) [inline test_requirements.txt:7]: Why do we need to pin?

Comment (valentin-krasontovitsch) [reply test_requirements.txt:7]: for some reason (i'm not sure why, i don't understand the inner working of pip orhow it resolves dependencies) the jenkins run was installing dash 2.4, and that one didn't have the method `wait_for_class_to_equal` that i needed for testing... so i just bumped it to a higher version.

but maybe, is there a notation like `>=2.5` that can be used? gonna google that real quick...

Comment (valentin-krasontovitsch) [reply test_requirements.txt]: fixed, it now says `>=2.5.1`

**Comment by valentin-krasontovitsch** [comment]
thank you : )

the currently latest screenshot displays the latest state - the test and version stuff didn't affect the layout, i hope ^^

so i'm gonna rebase, push, wait for tests to pass again, and then merge
